File: 2023/Day 8/perl/day8.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::Util qw(reduce);
use Math::BigInt;

sub create_paths {
    my @lines = @_;
    my %paths;
    foreach my $line (@lines) {
        my ($k, $l, $r) = $line =~ /[1-9A-Z]+/g;
        $paths{$k} = {"L" => $l, "R" => $r};
    }
    return %paths;
}
sub gcd {
    my ($a, $b) = @_;
    return $b == 0 ? $a : gcd($b, $a % $b);
}

sub lcm {
    my ($a, $b) = @_;
    return ($a * $b) / gcd($a, $b);
}

sub part1 {
    my ($paths, $dirs) = @_;
    my $pos = "AAA";
    my $solution = 0;
    while ($pos ne "ZZZ") {
        my $dir = $dirs->[$solution++ % scalar @$dirs];
        $pos = $paths->{$pos}{$dir};
    }
    return $solution;
}

sub part2 {
   my ($paths, $dirs) = @_;
    my @keys = grep { length($_) >= 3 && substr($_, 2, 1) eq "A" } keys %$paths;
    my @values = map {
        my $pos = $_;
        my $i = 0;
        while (substr($pos, 2, 1) ne "Z") {
            my $dir = $dirs->[$i++ % scalar @$dirs];
            $pos = $paths->{$pos}{$dir};
        }
        $i;
    } @keys;
    return reduce { lcm(Math::BigInt->new($a), Math::BigInt->new($b)) } 1, @values;
}

my $filename = "input.txt";
open my $fh, '<', $filename or die "Could not open file: $!";
chomp(my @data = <$fh>);
close $fh;
my @dirs = split //, $data[0]; 
my %paths = create_paths(@data[2..$#data]);
my $p1 = part1(\%paths, \@dirs);
my $p2 = part2(\%paths, \@dirs);
print "$p1\n";
print "$p2\n";

File: 2023/Day 8/ocaml/day8.ml
let read_file filename =
  let ic = open_in filename in
  let try_read () =
    try Some (input_line ic) with End_of_file -> None in
  let rec loop acc = match try_read () with
    | Some s -> loop (s :: acc)
    | None -> close_in ic; List.rev acc in
  loop []

let rec gcd a b = if b = 0 then a else gcd b (a mod b)
let lcm a b = a * b / (gcd a b)

let create_paths lines =
  let paths_table = Hashtbl.create 1024 in  (* Adjust size based on expected number of paths *)
  let parse_line line =
    try
      let parts = Str.split (Str.regexp "[=(), ]+") line in
      match parts with
      | key :: left :: right :: _ -> Some (key, [("L", left); ("R", right)])
      | _ -> None
    with Not_found -> None
  in
  List.iter (fun line ->
    match parse_line line with
    | Some (key, value) -> Hashtbl.add paths_table key value
    | None -> ()
  ) lines;
  paths_table

let rec part1 paths pos dirs solution =
  if pos = "ZZZ" then solution
  else
    let dir_index = solution mod Array.length dirs in
    let dir = Array.get dirs dir_index in
    let new_pos = Hashtbl.find paths pos in
    let next_pos = List.assoc dir new_pos in
    part1 paths next_pos dirs (solution + 1)

let part2 paths dirs =
  let keys = Hashtbl.fold (fun key _ acc -> key :: acc) paths [] in
  let filtered_keys = List.filter (fun k -> String.get k 2 = 'A') keys in
  let rec helper pos i =
    if String.get pos 2 = 'Z' then i
    else
      let dir_index = i mod Array.length dirs in
      let dir = Array.get dirs dir_index in
      let new_pos = Hashtbl.find paths pos in
      let next_pos = List.assoc dir new_pos in
      helper next_pos (i + 1)
  in
  let is = List.map (fun k -> helper k 0) filtered_keys in
  List.fold_left lcm 1 is


let () =
  let filename = Sys.argv.(1) in
  let input = read_file filename in
  let dirs_str = List.hd input in 
  let dirs = Array.of_list (List.init (String.length dirs_str) (fun i -> String.make 1 dirs_str.[i])) in
  let paths = create_paths (List.tl input) in
  Printf.printf "%d\n" (part1 paths "AAA" dirs 0);
  Printf.printf "%d\n" (part2 paths dirs);